httpbp/prometheus_test.go

```diff
@@ -0,0 +1,185 @@
+package httpbp
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"testing"
+
+	"github.com/reddit/baseplate.go"
+	"github.com/reddit/baseplate.go/ecinterface"
+	"github.com/reddit/baseplate.go/prometheusbp/promtest"
+)
+
+type exampleRequest struct {
+	Input string `json:"input"`
+}
+
+type exampleResponse struct {
+	Message string `json:"message"`
+}
+
+func TestPrometheusClientServerMetrics(t *testing.T) {
+	testCases := []struct {
+		name     string
+		code     string
+		success  string
+		method   string
+		endpoint string
+		size     int
+	}{
+		{
+			name:     "success get",
+			code:     "200",
+			success:  "true",
+			method:   http.MethodGet,
+			endpoint: "/test",
+		},
+		{
+			name:     "err post",
+			code:     "401",
+			success:  "false",
+			method:   http.MethodPost,
+			endpoint: "/error2",
+			size:     16,
+		},
+		{
+			name:     "internal err get",
+			code:     "500",
+			success:  "false",
+			method:   http.MethodGet,
+			endpoint: "/error",
+		},
+	}
+
+	const serverSlug = "testServer"
+
+	args := ServerArgs{
+		Baseplate: baseplate.NewTestBaseplate(baseplate.NewTestBaseplateArgs{
+			Config:          baseplate.Config{Addr: ":8080"},
+			EdgeContextImpl: ecinterface.Mock(),
+		}),
+		Endpoints: map[Pattern]Endpoint{
+			"/test": {
+				Name:    "test",
+				Methods: []string{http.MethodGet},
+				Handle:  func(ctx context.Context, w http.ResponseWriter, r *http.Request) error { return nil },
+			},
+			"/error2": {
+				Name:    "error",
+				Methods: []string{http.MethodPost},
+				Handle: func(ctx context.Context, w http.ResponseWriter, r *http.Request) error {
+					var req exampleRequest
+					if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
+						return fmt.Errorf("decoding %T: %w", req, err)
+					}
+					body := exampleResponse{
+						Message: fmt.Sprintf("Input: %q", req.Input),
+					}
+					return WriteJSON(w, Response{Body: body, Code: Unauthorized().code})
+				},
+			},
+			"/error": {
+				Name:    "error",
+				Methods: []string{http.MethodGet},
+				Handle: func(ctx context.Context, w http.ResponseWriter, r *http.Request) error {
+					return errors.New("test")
+				},
+			},
+		},
+		Middlewares: []Middleware{PrometheusServerMetrics(serverSlug)},
+	}
+
+	server, ts, err := NewTestBaseplateServer(args)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer server.Close()
+
+	middleware := PrometheusClientMetrics(serverSlug)
+	client := &http.Client{
+		Transport: middleware(http.DefaultTransport),
+	}
+
+	for _, tt := range testCases {
+		t.Run(tt.name, func(t *testing.T) {
+			serverLatency.Reset()
+			serverTotalRequests.Reset()
+			serverActiveRequests.Reset()
+			clientLatency.Reset()
+			clientTotalRequests.Reset()
+			clientActiveRequests.Reset()
+
+			serverTotalRequestLabels := []string{
```

Comment:
I'm starting to think we should swap the APIs to use prometheus.Labels, instead of a slice, since it's a real pain to compare these back and forth with the prometheus.go file and it either automatically detects or we can make it detect errors and give actionable failures. WDYT?

Reply:
good idea. i will make that change.